### PR TITLE
test: use lodash.isEqual instead of tap.strictSame

### PR DESCRIPTION
**Pull request by nechaido**
Current solution is a hack but it allows to see differences if needed.
Possible solutions:
* Use another testing framework.
* Patch **tap** to work properly with `NaN` (https://github.com/tapjs/node-tap/pull/371).
Fixes: #202.
Example output:
<details><summary> npm run test-todo</summary><p>

```
> metarhia-jstp@0.7.0-pre test-todo /home/nechaido/Projects/metarhia/jstp
> tap test/todo

test/todo/json5.js ................................ 179/194
  misc > readme-example
  not ok expected to not throw
    at:
      line: 64
      column: 48
      file: test/todo/json5.js
      function: test.doesNotThrow
    stack: |
      test.doesNotThrow (test/todo/json5.js:64:48)
      Test.test.test (test/todo/json5.js:64:16)
    type: TypeError
    source: |
      test.doesNotThrow(() => value = jstp.parse(file));
  
  numbers > hexadecimal-empty
  not ok expected to throw
    at:
      line: 86
      column: 16
      file: test/todo/json5.js
      type: Test
      function: test.test
    stack: |
      Test.test.test (test/todo/json5.js:86:16)
      fs.readdirSync.forEach (test/todo/json5.js:57:12)
      Array.forEach (native)
      Test.tap.test (test/todo/json5.js:47:35)
      testCases.forEach (test/todo/json5.js:46:7)
      Array.forEach (native)
      Object.<anonymous> (test/todo/json5.js:45:11)
    source: |
      test.throws(() => jstp.parse(file));
  
  numbers > infinity
  not ok expected to not throw
    at:
      line: 64
      column: 48
      file: test/todo/json5.js
      function: test.doesNotThrow
    stack: |
      test.doesNotThrow (test/todo/json5.js:64:48)
      Test.test.test (test/todo/json5.js:64:16)
    type: TypeError
    source: |
      test.doesNotThrow(() => value = jstp.parse(file));
  
  numbers > integer-with-float-exponent
  not ok expected to throw
    at:
      line: 86
      column: 16
      file: test/todo/json5.js
      type: Test
      function: test.test
    stack: |
      Test.test.test (test/todo/json5.js:86:16)
      fs.readdirSync.forEach (test/todo/json5.js:57:12)
      Array.forEach (native)
      Test.tap.test (test/todo/json5.js:47:35)
      testCases.forEach (test/todo/json5.js:46:7)
      Array.forEach (native)
      Object.<anonymous> (test/todo/json5.js:45:11)
    source: |
      test.throws(() => jstp.parse(file));
  
  numbers > integer-with-hexadecimal-exponent
  not ok expected to throw
    at:
      line: 86
      column: 16
      file: test/todo/json5.js
      type: Test
      function: test.test
    stack: |
      Test.test.test (test/todo/json5.js:86:16)
      fs.readdirSync.forEach (test/todo/json5.js:57:12)
      Array.forEach (native)
      Test.tap.test (test/todo/json5.js:47:35)
      testCases.forEach (test/todo/json5.js:46:7)
      Array.forEach (native)
      Object.<anonymous> (test/todo/json5.js:45:11)
    source: |
      test.throws(() => jstp.parse(file));
  
  numbers > integer-with-negative-float-exponent
  not ok expected to throw
    at:
      line: 86
      column: 16
      file: test/todo/json5.js
      type: Test
      function: test.test
    stack: |
      Test.test.test (test/todo/json5.js:86:16)
      fs.readdirSync.forEach (test/todo/json5.js:57:12)
      Array.forEach (native)
      Test.tap.test (test/todo/json5.js:47:35)
      testCases.forEach (test/todo/json5.js:46:7)
      Array.forEach (native)
      Object.<anonymous> (test/todo/json5.js:45:11)
    source: |
      test.throws(() => jstp.parse(file));
  
  numbers > integer-with-negative-hexadecimal-exponent
  not ok expected to throw
    at:
      line: 86
      column: 16
      file: test/todo/json5.js
      type: Test
      function: test.test
    stack: |
      Test.test.test (test/todo/json5.js:86:16)
      fs.readdirSync.forEach (test/todo/json5.js:57:12)
      Array.forEach (native)
      Test.tap.test (test/todo/json5.js:47:35)
      testCases.forEach (test/todo/json5.js:46:7)
      Array.forEach (native)
      Object.<anonymous> (test/todo/json5.js:45:11)
    source: |
      test.throws(() => jstp.parse(file));
  
  numbers > integer-with-positive-float-exponent
  not ok expected to throw
    at:
      line: 86
      column: 16
      file: test/todo/json5.js
      type: Test
      function: test.test
    stack: |
      Test.test.test (test/todo/json5.js:86:16)
      fs.readdirSync.forEach (test/todo/json5.js:57:12)
      Array.forEach (native)
      Test.tap.test (test/todo/json5.js:47:35)
      testCases.forEach (test/todo/json5.js:46:7)
      Array.forEach (native)
      Object.<anonymous> (test/todo/json5.js:45:11)
    source: |
      test.throws(() => jstp.parse(file));
  
  numbers > integer-with-positive-hexadecimal-exponent
  not ok expected to throw
    at:
      line: 86
      column: 16
      file: test/todo/json5.js
      type: Test
      function: test.test
    stack: |
      Test.test.test (test/todo/json5.js:86:16)
      fs.readdirSync.forEach (test/todo/json5.js:57:12)
      Array.forEach (native)
      Test.tap.test (test/todo/json5.js:47:35)
      testCases.forEach (test/todo/json5.js:46:7)
      Array.forEach (native)
      Object.<anonymous> (test/todo/json5.js:45:11)
    source: |
      test.throws(() => jstp.parse(file));
  
  numbers > lone-decimal-point
  not ok expected to throw
    at:
      line: 86
      column: 16
      file: test/todo/json5.js
      type: Test
      function: test.test
    stack: |
      Test.test.test (test/todo/json5.js:86:16)
      fs.readdirSync.forEach (test/todo/json5.js:57:12)
      Array.forEach (native)
      Test.tap.test (test/todo/json5.js:47:35)
      testCases.forEach (test/todo/json5.js:46:7)
      Array.forEach (native)
      Object.<anonymous> (test/todo/json5.js:45:11)
    source: |
      test.throws(() => jstp.parse(file));
  
  numbers > nan
  not ok expected to not throw
    at:
      line: 64
      column: 48
      file: test/todo/json5.js
      function: test.doesNotThrow
    stack: |
      test.doesNotThrow (test/todo/json5.js:64:48)
      Test.test.test (test/todo/json5.js:64:16)
    type: TypeError
    source: |
      test.doesNotThrow(() => value = jstp.parse(file));
  
  numbers > noctal-with-leading-octal-digit
  not ok expected to not throw
    at:
      line: 75
      column: 50
      file: test/todo/json5.js
      function: test.doesNotThrow
    stack: |
      test.doesNotThrow (test/todo/json5.js:75:50)
      Test.test.test (test/todo/json5.js:75:18)
    type: SyntaxError
    source: |
      test.doesNotThrow(() => value = jstp.parse(file));
  
  objects > unquoted-keys
  not ok expected to not throw
    at:
      line: 64
      column: 48
      file: test/todo/json5.js
      function: test.doesNotThrow
    stack: |
      test.doesNotThrow (test/todo/json5.js:64:48)
      Test.test.test (test/todo/json5.js:64:16)
    type: SyntaxError
    source: |
      test.doesNotThrow(() => value = jstp.parse(file));
  
  todo > unicode-escaped-unquoted-key
  not ok expected to not throw
    at:
      line: 64
      column: 48
      file: test/todo/json5.js
      function: test.doesNotThrow
    stack: |
      test.doesNotThrow (test/todo/json5.js:64:48)
      Test.test.test (test/todo/json5.js:64:16)
    type: SyntaxError
    source: |
      test.doesNotThrow(() => value = jstp.parse(file));
  
  todo > unicode-unquoted-key
  not ok expected to not throw
    at:
      line: 64
      column: 48
      file: test/todo/json5.js
      function: test.doesNotThrow
    stack: |
      test.doesNotThrow (test/todo/json5.js:64:48)
      Test.test.test (test/todo/json5.js:64:16)
    type: SyntaxError
    source: |
      test.doesNotThrow(() => value = jstp.parse(file));

total ............................................. 179/194
```

### CODE REVIEW

**Comment by nechaido** [comment]
@belochub ping

**Comment by belochub** [comment]
@nechaido, you told me in personal conversation that it wasn't ready.

**Comment by nechaido** [comment]
@belochub later that day I said it was ready. that is way a pinged you.

File: test/todo/json5.js [hunk 1]
```diff
@@ -53,25 +55,35 @@ testCases.forEach((testCase) => {
...
-            const supportedTests = supportedByUs[testCase.name];
-            if (supportedTests && supportedTests.includes(testName)) {
-              test.strictSame(jstp.parse(file), extendedEval(file));
-            } else {
-              test.throws(() => jstp.parse(file));
-            }
-            break;
+        if (ext === '.json') {
+          let value;
+          test.doesNotThrow(() => value = jstp.parse(file));
```
Comment (belochub) [inline test/todo/json5.js]: This testing is redundant and hides error message in case when parsing actually throws.

**Comment by aqrln** [comment]
Ugh... I understand the reasons behind this change, but do we *really* need it? I'd rather not introduce such a dirty hack if there's a solution (I believe there is, but lmk if I'm wrong!) that will result in the tests passing or failing correctly, but maybe with a bit inconvenient output when they fail. Tests shouldn't fail consistently, and when they do, inserting a `debugger` statement before an assertion and running `node --inspect test.js` is cheap.

At this point I'm -1 on this.

**Comment by aqrln** [comment]
Just to be clear: the main idea of the PR looks fine. Using `tap.ok(lodash.isEqual(...))` instead of `tap.strictSame(...)` and leaving a comment explaining the reason of doing so would be a legit workaround. My -1 applies to all the other changes in this PR.

**Comment by nechaido** [comment]
@belochub removed.

**Comment by nechaido** [comment]
@aqrln I don't like this hack either, so I say it's for @belochub to decide.
The only thing I can propose is to have `test-todo` and `test-todo-verbose`, first one is only for checking that everything is ok, when the second shows what is the problem using hack introduced before, this way nobody will need to insert debug logs manually.

File: test/todo/json5.js [hunk 2]
```diff
@@ -62,6 +62,8 @@ testCases.forEach((testCase) => {
           case '.json5': {
             const value = jstp.parse(file);
             const expected = extendedEval(file);
+            // Currently tap.strictSame() doesn't work well with NaNs.
+            // That is why we check if lodash.isequal return true.
             test.assert(deepEqual(value, expected));
             if (!test.passing())
               console.worn(difference(value, expected));
```
Comment (aqrln) [inline test/todo/json5.js]: s/worn/warn

File: test/todo/json5.js [hunk 3]
```diff
@@ -57,13 +59,26 @@ testCases.forEach((testCase) => {
           case '.json':
             test.strictSame(jstp.parse(file), JSON.parse(file));
             break;
-          case '.json5':
-            test.strictSame(jstp.parse(file), extendedEval(file));
+          case '.json5': {
+            const value = jstp.parse(file);
+            const expected = extendedEval(file);
+            // Currently tap.strictSame() doesn't work well with NaNs.
+            // That is why we check if lodash.isequal return true.
```
Comment (aqrln) [inline test/todo/json5.js]: s/return/returns

**Comment by nechaido** [comment]
@belochub @aqrln fixed

**Comment by belochub** [comment]
Landed in https://github.com/metarhia/jstp/commit/fb1dbef9a4cd3c7ffe7e81aa792669686149ec82.

**Pull request closed without merging**